### PR TITLE
adds anyhow where result used to be

### LINKED ISSUE #851: Make error handling more consistent throughout `forc`
Currently there's a combination of `anyhow` and using `Result<T, String>`, with the latter being more prevalent.

We should switch to using `anyhow` everywhere, or consider using `thiserror` seeing as we currently also expose `forc` as a library.

### PR DESCRIPTION
**Pull request by eureka-cpu**
Intended to close #851

I've gone through and replaced any instance I could find of `Result<(), String>` with `anyhow`, along with error messages.

### CODE REVIEW

File: forc/src/utils/restricted_names.rs [hunk 1]
```diff
@@ -93,36 +94,34 @@ pub fn is_glob_pattern<T: AsRef<str>>(name: T) -> bool {
     name.as_ref().contains(&['*', '?', '[', ']'][..])
 }
 
-#[test]
-fn test_invalid_char() {
-    assert_eq!(
-        contains_invalid_char("test#proj", "package name"),
```
Comment (JoshuaBatty) [inline forc/src/utils/restricted_names.rs:99]: You should be able to get this to work again by just mapping the error returned by anyhow back to a string

`contains_invalid_char("test#proj", "package name").map_err(|e| e.to_string())`

File: forc/src/utils/restricted_names.rs [hunk 2]
```diff
@@ -93,36 +94,34 @@ pub fn is_glob_pattern<T: AsRef<str>>(name: T) -> bool {
...
-        std::result::Result::Err(
-            "invalid character `#` in package name: `test#proj`, \
-                characters must be Unicode XID characters \
-                (numbers, `-`, `_`, or most letters)"
-                .into()
-        )
-    );
-    assert_eq!(
-        contains_invalid_char("test proj", "package name"),
```
Comment (JoshuaBatty) [inline forc/src/utils/restricted_names.rs:108]: same here

`contains_invalid_char("", "package name").map_err(|e| e.to_string())`

File: forc/src/utils/restricted_names.rs [hunk 3]
```diff
@@ -93,36 +94,34 @@ pub fn is_glob_pattern<T: AsRef<str>>(name: T) -> bool {
...
-        std::result::Result::Err(
-            "invalid character ` ` in package name: `test proj`, \
-                characters must be Unicode XID characters \
-                (numbers, `-`, `_`, or most letters)"
-                .into()
-        )
-    );
-    assert_eq!(
-        contains_invalid_char("", "package name"),
```
Comment (JoshuaBatty) [inline forc/src/utils/restricted_names.rs:117]: and here 
`contains_invalid_char("", "package name").map_err(|e| e.to_string())`

File: forc/src/utils/restricted_names.rs [hunk 4]
```diff
@@ -93,36 +94,34 @@ pub fn is_glob_pattern<T: AsRef<str>>(name: T) -> bool {
...
-        std::result::Result::Err(
-            "package name cannot be left empty, \
-            please use a valid name"
-                .into()
-        )
-    );
-    assert_eq!(
-        contains_invalid_char("test_proj", "package name"),
```
Comment (JoshuaBatty) [inline forc/src/utils/restricted_names.rs:125]: `contains_invalid_char("test_proj", "package name").map_err(|e| e.to_string())`

File: forc/src/utils/restricted_names.rs [hunk 5]
```diff
@@ -93,36 +94,34 @@ pub fn is_glob_pattern<T: AsRef<str>>(name: T) -> bool {
...
-        std::result::Result::Ok(())
-    );
-}
+// unfortunately there's not a good way to test after converting to `anyhow`,
```
Comment (JoshuaBatty) [inline forc/src/utils/restricted_names.rs:97]: Can get rid of this whole commented block now

**Review by JoshuaBatty — CHANGES_REQUESTED** [review]
Nice, looks good thanks. Just left some comments to help you get the tests working again. 

There does still remain the question if we should be considering using `thiserror` instead of `anyhow` if we are exposing `forc` as a library. But this is still an upgrade from the current error handling and can be addressed in a separate PR if we decide to move to `thiserror`.

**Comment by mitchmindtree** [comment]
Ahh was hoping to have this addressed after landing #825 but I should have mentioned so in the issue :smiling_face_with_tear: if this happens to land first I can rebase on top no worries :+1: 

---

I would like to see `thiserror` with more granular types eventually, but I think separating the parts of `forc` that we wish to expose as a library more clearly first might help to go about this, I'll open an issue with some thoughts on this. For now though I think `anyhow` and this PR makes a lot of sense!

**Comment by JoshuaBatty** [comment]
I think it probably looked tidier the way you first had it, eg: 

```
assert_eq!(
    contains_invalid_char("test#proj", "package name").map_err(|e| e.to_string()),
    std::result::Result::Err(
        "invalid character `#` in package name: `test#proj`, \
            characters must be Unicode XID characters \
            (numbers, `-`, `_`, or most letters)"
            .into()
    )
);
```

**Review by JoshuaBatty — APPROVED** [review]
LGTM, thanks for addressing those changes.

**Pull request merged**